stdlib/public/core/Sequence.swift

```diff
@@ -950,13 +962,32 @@ extension Sequence {
   ///
   /// - Parameter body: A closure that takes an element of the sequence as a
   ///   parameter.
+  @warn_unused_result
   public func forEach(
     _ body: @noescape (Iterator.Element) throws -> Void
   ) rethrows {
     for element in self {
       try body(element)
     }
   }
+  
+  /// Returns the first element of the sequence that satisfies the given
+  /// predicate or nil if no such element is found.
+  ///
+  /// - Parameter predicate: A closure that takes an element of the
+  ///   sequence as its argument and returns a Boolean value indicating
+  ///   whether the element is a match.
+  /// - Returns: The first match or `nil` if there was no match.
+  public func first(
+    predicate: @noescape (Iterator.Element) throws -> Bool
+  ) rethrows -> Iterator.Element? {
+    for element in self {
```

Comment:
This didn't work at some point in the past but rethrow has gotten smarter since then. I tested that it doesn't interfere with throwing custom errors, nor does it break rethrows for a non-throwing closure. 

I wanted to get your opinion on this before I push a commit since it looks "interesting" at first glance 😁



```swift
enum _ForEachStopIteration: ErrorProtocol {
    case stopIteration
}

extension Sequence {
    /// Calls the given closure on each element in the sequence in the same order
    /// as a `for`-`in` loop.
    ///
    /// - Parameter body: A closure that takes an element of the sequence as a
    ///   parameter. Return `true` from this closure to stop iteration.
    ///   Returning `false` will continue iteration.
    func _forEach(_ body: @noescape (Iterator.Element) throws -> Bool) rethrows {
        do {
            try self.forEach {
                if try body($0) {
                    throw _ForEachStopIteration.stopIteration
                }
            }
        } catch is _ForEachStopIteration { }
    }

    /// <insert find description here>
    public func find(where predicate: (Iterator.Element) throws -> Bool) rethrows -> Iterator.Element? {
        var foundElement: Iterator.Element? = nil
        try self._forEach {
            if try predicate($0) {
                foundElement = $0
                return true
            }
            return false
        }
        return foundElement
    }
}
```